torch/utils/data/dataloader.py

```diff
@@ -87,32 +111,33 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
         random.seed(seed)
         torch.manual_seed(seed)
 
-        # Do not wait for putting thread to join when this worker exits.
-        # Otherwise, this worker may always be waiting to put and doesn't check
-        # index_queue and done_event for termination signal.
         data_queue.cancel_join_thread()
 
         if init_fn is not None:
             init_fn(worker_id)
 
         watchdog = ManagerWatchdog()
 
-        while True:
+        while watchdog.is_alive():
             try:
-                r = index_queue.get(timeout=MANAGER_STATUS_CHECK_INTERVAL)
+                r = index_queue.get(timeout=MP_STATUS_CHECK_INTERVAL)
             except queue.Empty:
-                if watchdog.is_alive() and not done_event.is_set():
-                    continue
-                else:
-                    break
-            # use done_event so that we can get faster exiting signal even if there
-            # are still indices in index_queue
-            if r is None or done_event.is_set():
-                break
+                continue
```

Comment:
I don't think this is equivalent to the code we used previously. We used to not only check that the watchdog is alive, but also that the done event is not set, before we were allowed to continue an iteration. So theoretically a situation where the manager stays alive, but never pushes to a queue again, would put you in an infinite busy loop in here. That's not very nice, but I guess it should never happen, since our invariant is that everyone will eventually get their `None` signal at the end...

Reply:
Yes. The invariant says that if main process is alive, the workers will get a `None`, eventually. We must get that `None` even when `done_event` is set, because otherwise main process putter thread may hang.